Add initial version of new documentation (still incomplete) (#22)
rwth-i6/rasr: doc/architecture.rst

Architecture
============

> Build environment

Author: But let's do it in a separate PR :grinning:

It can be used for various purpose:

* ``PythonFeatureScorer`` (anywhere for feature scoring, such as decoding or generating alignments)
* ``PythonTrainer`` (interface for the legacy RASR NN training code, also used in decoding together with TrainerFeatureScorer)
* ``PythonSegmentOrdering`` (sorting or filtering the dataset)
* ``PythonLayer`` (plug in for the legacy RASR NN code, decoding/training/anything)
* ``PythonControl`` (very custom scriptable control to e.g. get data out of dataset, or calculate soft/hard alignments, generate FSAs, etc.)
* :doc:`training/sequence_training.rst` (RASR gets the posteriors and returns loss and error signal)

On RETURNN side, several implementations for these interfaces exists:

* `SprintDataset <https://github.com/rwth-i6/returnn/blob/master/SprintDataset.py>`_ and `ExternSprintDataset <https://github.com/rwth-i6/returnn/blob/master/SprintExternInterface.py>`_. (Which can either use PythonTrainer or PythonControl as the RASR entry point.)
* Sequence training
* `Full-sum training <https://www-i6.informatik.rwth-aachen.de/publications/download/1035/Zeyer--2017.pdf>`_ (gets FSA via PythonControl)

> Note that there is also the Tensorflow module which provides an own FeatureScorer.

Reviewer: Fix Denglish
```suggestion
Note that there is also the Tensorflow module which provides a separate FeatureScorer.
```

MODULE_TENSORFLOW
^^^^^^^^^^^^^^^^^

In order to use TF C++ API, we have to be able to include TF headers (e.g. tensorflow/core/framework/tensor.h) and link against libtensorflow_cc.so and libtensorflow_framework.so. Also, we want TF to include Intel MKL support and link against libiomp5.so and libmklml_intel.so. The only way to get libtensorflow_cc.so is to build TF from sources.

Dependencies
^^^^^^^^^^^^

You can run ``scripts/requirements.sh`` to verify if your build environment meets the requirements. As of January 2020, RASR is known to compile with gcc between 4.8 and 7.3, optionally with Python 3.6 and TF 1.12 to 1.15. 

Currently, RASR requires C++11. As of January 2020, we plan to upgrade to C++14 or 17 at some later point in time because of some internal dependencies in our infrastructure.

> Please note that for the streaming recognizer, the set of dependencies is larger (gRPC, Protobuf, Boost, streaming API, OPUS resampling library).

Reviewer: Information refer to AppTek IP. Streaming recognizer is not part of the github rasr version

**TODO**: discuss 

* minus log prob domain
* buffered access
* reset()
* finalize()

**Gaussian mixture models**

GMMs are configured via selector ``mixture-set``.

**Legacy DNN**

Feed forward DNNs (as used e.g. in segmenter) are configured via selector ``neural-network`` (Nn/NeuralNetwork.cc).

**Tensorflow models**

TF models are configured via selector ``loader`` (Tensorflow/TensorflowForwardNode.cc, Tensorflow/TensorflowFeatureScorer.cc) and have to specify the parameters

* type - meta or vanilla
* meta-graph-file - meta file with the TF graph
* saved-model-file - file name prefix that can be expanded to .data and .index that contain the actual parameters
* required-libraries - colon-separated list of shared libraries that are loaded via `tf::Env::Default->LoadLibrary <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/env.h#L314>`_, e.g. for loading pre-compiled custom TF ops like LSTM kernels.

Alignment
^^^^^^^^^

See :ref:`Alignment`, :doc:`training/alignment_generation.rst` and :doc:`training/converting_alignment_formats.rst`

> The alignment is a frame-level mapping from feature vector index within a segment (no time stamps are stored here, just integer indices 0, 1, 2, ... relative to segment start) to a triphone HMM state. Please note that the alignment always stores triphones, even if the user is only interested in monophones. Also, we don't store tied states (although we support this feature) so that we can apply any state tying strategy on top of an existing alignment. Alignment format can store frame weights (float), e.g. for sophisticated AM training. The alignment stores both, the allophone index and the state index in one 32-bit integer: the first 26 bits of the allphone state id are used for the allophone index; the latter 6 bits for the state. This means that the maximal id for the allophone index is 2^26 and for HMM state is 2^6 (we mostly use 3 in ASR and a little bit more in sign language or handwriting recognition).

Reviewer: > the first 26 bits of the allphone state id are used for the allophone index; the latter 6 bits for the state.

I think the most significant bits store the state id and the least significant bits the allophone index.

Reviewer: My comment was not referring to the nomenclature "first n bits" vs "most significant bits" etc, but because I think there is a factual error here:
The text now states that the _least_ significant bits store the state whereas I have in mind that the state is stored in the _most_ significant bits.

Can you please verify/clarify that the text is correct and my recollection is wrong?

Author: According to rasr_cache.py your memory is correct. I have updated the documentation.

Please note that no allophones are needed during recognition, such that both ``add-from-lexicon`` and ``add-all`` should be set to false.

Cache
^^^^^
See :ref:`Archive`

RASR uses its own data structures for features, alignments, lattices etc. The usual storage format is an archive (sometimes also called cache), which is a binary format that supports compression. The available formats inherit from Core::Archive (class names = file names in src/Core)

* ``Core::BundleArchive`` - see :ref:`Bundle Archive`
* ``Core::DirectoryArchive`` - rarely used
* ``Core::FileArchive`` - used for features, alignments, lattices etc.

A ``FileArchive`` can be considered a "tarball" that can hold multiple independent files. Its format is specified in ``src/Core/FileArchive.cc`` and can be read using the :ref:`Archiver Tool` or `SprintCache.py <https://github.com/rwth-i6/returnn/blob/master/SprintCache.py>`_. See also ``src/Tools/Archiver/Archiver.cc`` for example usage. The low level I/O is implemented in ``Core::BinaryStream``.

The zlib compression is specified in ``src/Core/Archive.cc``

> Channel

Reviewer: duplicate information with channel.rst

The :ref:`acoustic model <Acoustic model>` owns a ``Am::ClassicStateModel`` and a ``Am::ClassicHmmTopologySet``. The latter defines the number of states per phoneme and whether or not the across word modeling is enabled. There is also support for "sub-states", which is an artificial duplication of HMM states:

* with 3 states per phone and ''no'' repetition, the state sequence for a phoneme is ``s1 s2 s3``
* with repetition enabled (``hmm.state-repetitions=2``), the state sequence becomes ``s1 s1' s2 s2' s3 s3'`` with all valid transitions defined as usual, but

  * ``p(x|si') := p(x|si)  for i = 1, 2, 3``
  * no new emission classes are introduced

Please note how "repetition" is a misnomer, because ``hmm.state-repetitions=1`` means "only one state instance, no additional repetition".

Also, the AM owns a ``Am::TransitionModel``, which holds a vector of ``Am::StateTransitionModel``, one for each state (depending on the value of ``hmm.tying-type``). The transition model can execute ``TransitionModel::apply()`` on an FST to add loop and skip
transitions to a "flat" automaton (meaning that it does not
contain loops and skips).

> The emission labels, i.e. the labels

Author: I would leave it in anyway. Many people never open the source code to read the documentation in there.
